instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt

```diff
@@ -113,11 +116,150 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
+            mockLocationReplayerRule.playRoute(routes.first())
+            mapboxNavigation.startTripSession()
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(routes)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation, true)
+        firstAlternative.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("alternatives failed") {
+            Espresso.onIdle()
+        }
+        firstAlternative.unregister()
+        assertTrue(firstAlternative.calledOnMainThread)
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
         // Verify alternative routes events were triggered.
         assertEquals(2, routes.size)
         assertTrue(mapboxNavigation.getRoutes().isNotEmpty())
         assertNotNull(firstAlternative.routeProgress)
         assertNotNull(firstAlternative.alternatives)
+
+        firstAlternative.alternatives!!.forEach { alternativeRoute ->
+            routes.forEach { originalRoute ->
+                assertNotEquals(originalRoute, alternativeRoute)
+            }
+        }
+    }
+
+    /**
+     * The test verifies that if set routes to Navigation that are come from alternatives + n
+     * (where n >= 1) external routes, alternatives routes observer is not triggered with these routes.
+     * That was happening in legacy versions of NN.
+     */
+    @Test
+    fun additional_alternative_is_not_force_to_invoke_alternatives_observer() {
+        // Prepare with alternative routes.
+        setupMockRequestHandlers(coordinates)
+        val primaryRouteOnly = listOf(requestDirectionsRouteSync(coordinates).first())
+
+        // Play the slower alternative route.
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
+            mockLocationReplayerRule.playRoute(primaryRouteOnly.first())
+            mapboxNavigation.startTripSession()
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(primaryRouteOnly)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation, true)
```

Comment:
how does NN get alternatives during the tests? As far as I understand it will wait until you passe the deviation point and then do a route request. But there is no setup responses for this case in mocked web server 🤔 
How does it work?

Reply:
it works independently, so alternatives are produced by NN for the current mocked route

Reply:
The test is agnostic to the type of route, we don't need to know it. But NN might produce routes by onboard or offboard, depending on current conditions (like internet connection availability).

Reply:
Sorry, didn't get this 🙂 
Let's discuss on a call later